url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8/src/com/fuzzylite/imex/FllImporter.java

```diff
@@ -0,0 +1,340 @@
+                            + "expected a colon at line " + lineNumber + ": " + line);
+                }
+                String key = line.substring(0, colon).trim();
+                String value = line.substring(colon + 1).trim();
+                if ("Engine".equals(key)) {
+                    engine.setName(value);
+                    continue;
+                } else {
+                    processPending = ("InputVariable".equals(key)
+                            || "OutputVariable".equals(key)
+                            || "RuleBlock".equals(key));
+                }
+                if (processPending) {
+                    process(tag, block.toString(), engine);
+                    block.setLength(0);
+                    processPending = false;
+                    tag = key;
+                }
+                block.append(String.format("%s:%s\n", key, value));
+            }
+            process(tag, block.toString(), engine);
+            reader.close();
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
+        }
+        return engine;
+    }
+
+    protected void process(String tag, String block, Engine engine) throws Exception {
+        if (tag.isEmpty()) {
+            return;
+        }
+        if ("InputVariable".equals(tag)) {
+            processInputVariable(block, engine);
+        } else if ("OutputVariable".equals(tag)) {
+            processOutputVariable(block, engine);
+        } else if ("RuleBlock".equals(tag)) {
+            processRuleBlock(block, engine);
+        } else {
+            throw new RuntimeException("[import error] block tag <" + tag + "> "
+                    + "not recognized");
+        }
+    }
+
+    protected void processInputVariable(String block, Engine engine) throws Exception {
+        BufferedReader reader = new BufferedReader(new StringReader(block));
+        String line;
+        InputVariable inputVariable = new InputVariable();
+        engine.addInputVariable(inputVariable);
+        while ((line = reader.readLine()) != null) {
+            Op.Pair<String, String> keyValue = parseKeyValue(line, ':');
+            if ("InputVariable".equals(keyValue.first)) {
+                inputVariable.setName(keyValue.second);
+            } else if ("enabled".equals(keyValue.first)) {
+                inputVariable.setEnabled(parseBoolean(keyValue.second));
+            } else if ("range".equals(keyValue.first)) {
+                Op.Pair<Double, Double> range = parseRange(keyValue.second);
+                inputVariable.setRange(range.first, range.second);
+            } else if ("term".equals(keyValue.first)) {
+                inputVariable.addTerm(parseTerm(keyValue.second, engine));
+            } else {
+                throw new RuntimeException("[import error] "
+                        + "key <" + keyValue.first + "> " + "not recognized in pair <"
+                        + Op.join(":", keyValue.first, keyValue.second) + ">");
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 155:161, 190:196.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.